File: Resources/Packet/Hit.cs
using System.IO;
using Resources.Utilities;

namespace Resources.Packet {
    public class Hit {
        public const int packetID = 7;

        public long attacker;
        public long target;
        public float damage;
        public int critical;//bool?
        public int stuntime;
        public int paddingA;
        public LongVector position = new LongVector();
        public FloatVector direction = new FloatVector();
        public byte skill;
        public byte type;
        public byte showlight;
        public byte paddingB;

        public Hit() { }

        public Hit(BinaryReader reader) {
            attacker = reader.ReadInt64();
            target = reader.ReadInt64();
            damage = reader.ReadSingle();
            critical = reader.ReadInt32();
            stuntime = reader.ReadInt32();
            paddingA = reader.ReadInt32();
            position = new LongVector(reader);
            direction = new FloatVector(reader);
            skill = reader.ReadByte();
            type = reader.ReadByte();
            showlight = reader.ReadByte();
            paddingB = reader.ReadByte();
        }

        public void Write(BinaryWriter writer, bool writePacketID = true) {
            if(writePacketID) {
                writer.Write(packetID);
            }
            writer.Write(attacker);
            writer.Write(target);
            writer.Write(damage);
            writer.Write(critical);
            writer.Write(stuntime);
            writer.Write(paddingA);
            position.Write(writer);
            direction.Write(writer);
            writer.Write(skill);
            writer.Write(type);
            writer.Write(showlight);
            writer.Write(paddingB);
        }
    }
}

File: common/src/main/kotlin/packets/Hit.kt
package packets

import utils.*

data class Hit(
	val attacker: Long,
	val target: Long,
	val damage: Float,
	val critical: Boolean,
	val stuntime: Int,
	val paddingA: Int,
	val position: Vector3<Long>,
	val direction: Vector3<Float>,
	val isYellow: Boolean,
	val damageType: DamageType,
	val flash: Boolean,
	val paddingB: Byte
) : Packet(Opcode.Hit), SubPacket {
	override suspend fun writeTo(writer: Writer) {
		writer.writeLong(attacker)
		writer.writeLong(target)
		writer.writeFloat(damage)
		writer.writeBoolean(critical); writer.pad(3)
		writer.writeInt(stuntime)
		writer.writeInt(paddingA)
		writer.writeVector3Long(position)
		writer.writeVector3Float(direction)
		writer.writeBoolean(isYellow)
		writer.writeByte(damageType.value)
		writer.writeBoolean(flash)
		writer.writeByte(paddingB)
	}

	companion object {
		suspend fun readFrom(reader: Reader): Hit {
			return Hit(
				attacker = reader.readLong(),
				target = reader.readLong(),
				damage = reader.readFloat(),
				critical = reader.readInt() > 0,
				stuntime = reader.readInt(),
				paddingA = reader.readInt(),
				position = reader.readVector3Long(),
				direction = reader.readVector3Float(),
				isYellow = reader.readBoolean(),
				damageType = DamageType(reader.readByte()),
				flash = reader.readBoolean(),
				paddingB = reader.readByte()
			)
		}
	}
}

inline class DamageType(val value: Byte) {
	companion object {
		val Block = DamageType(1)
		val Normal = DamageType(2)
		val Miss = DamageType(3)
		val Invisible = DamageType(4)
		val Absorb = DamageType(5)
		val Invisible2 = DamageType(6)
	}
}